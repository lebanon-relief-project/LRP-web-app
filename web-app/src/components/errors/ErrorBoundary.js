import React from "react";
import ErrorPage from "../../pages/errorPage";

export default class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { error: null, errorInfo: null };

    props.history &&
      props.history.listen((location, action) => {
        this.setState({ error: null, errorInfo: null });
      });
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      error: true,
      errorInfo: errorInfo,
    });
  }

  render() {
    if (this.state.error) {
      return <ErrorPage />;
    }

    return this.props.children;
  }
}
