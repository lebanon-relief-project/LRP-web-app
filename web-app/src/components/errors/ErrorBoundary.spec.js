import React from "react";
import { render } from "@testing-library/react";
import "jest-styled-components";
import ErrorBoundary from "./ErrorBoundary";

jest.mock("../../pages/errorPage", (props) => {
  return (props) => <div>Error page</div>;
});

beforeEach(() => {
  jest.spyOn(console, "error").mockImplementation(() => {});
});

describe("the ErrorPage component", () => {
  it("should render children when no error happenned", () => {
    let Child = () => {
      return <div>Regular child</div>;
    };
    const { container, getByText } = render(
      <ErrorBoundary>
        <Child />
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
    expect(getByText("Regular child")).toBeTruthy();
  });

  it("should render Error page when exception occurs", () => {
    let Child = () => {
      throw new Error("ugly error in component");
    };

    const { container, queryByText } = render(
      <ErrorBoundary>
        <Child />
      </ErrorBoundary>
    );

    expect(container).toMatchSnapshot();
    expect(queryByText("Regular child")).toBeFalsy();
    expect(queryByText("Error page")).toBeTruthy();
  });
});
