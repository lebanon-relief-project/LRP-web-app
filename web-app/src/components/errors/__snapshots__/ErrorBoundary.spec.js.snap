// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ErrorPage component should render Error page when exception occurs 1`] = `
<div>
  <div>
    Error page
  </div>
</div>
`;

exports[`the ErrorPage component should render children when no error happenned 1`] = `
<div>
  <div>
    Regular child
  </div>
</div>
`;
